WIP Briefing Refactor - Adds jinja2 dependency

walterroach opened the pull request:
Complete rework of briefinggen module to try and make it more maintainable.  

Uses jinja2 for templating.

Creates a unified interface for kneeboards and briefing on the Operation class.  This may be better done by extending the pydcs Mission class, and generating the kneeboards and briefings as a pre-save hook on .miz save- but for now I've reworked it where it was being generated before.

Check my changes to qt_ui/windows/basemenu/ground_forces/QGroundForcesStrategySelector.py
When "Defensive" was selected, it would set the frontline CombatStance to an instance of the CombatStance.DEFENSIVE enum, however for all other strategies it was setting strategy to the int value.  I searched and could not find any place where this is not appropriate, but it's possible I missed something.

WalterHaynieChq committed:
parent f03121af5ae214e7a58f37e4e1d4ad9102a2f35e

first pass briefing refactor

briefing fixes

briefing fixes

Stop briefing generate being called twice

Stop frontline advantage string being appended
when there are no units.

jinja template

always return enum instance in Strategy Selector

For some reason on DEFENSE, enum is appended to control point stance,
but on all other the enum.value is added instead.

I don't see any case where the value is used, but there are many
cases that the enum instance is evaluated against.

type issue

junja's not a thing

swap mapping with dict

jinja template

always return enum instance in Strategy Selector

For some reason on DEFENSE, enum is appended to control point stance,
but on all other the enum.value is added instead.

I don't see any case where the value is used, but there are many
cases that the enum instance is evaluated against.

type issue

Update build.yml

junja's not a thing

swap mapping with dict

Restore build job

WalterHaynieChq committed:
linting and formatting

walterroach committed:
refactor frontline situation briefing

walterroach committed:
Properly get enum instead of value on strategy
selecter

walterroach committed:
var change

walterroach committed:
refactor frontline brief generation

walterroach committed:
working briefinggen refactor

WalterHaynieChq committed:
cleanup

WalterHaynieChq committed:
Move kneeboard and briefing gen to unified
interface

WalterHaynieChq committed:
docstrings and cleanup

WalterHaynieChq committed:
init args change

WalterHaynieChq committed:
template cleanup

walterroach commented:
Known issue- the departure delay times are wrong (they're wrong in the dev branch too).

DanAlbert commented on game/operation/operation.py:
```diff
@@ -76,6 +75,7 @@ def __init__(self,
         self.to_cp = to_cp
         self.is_quick = False
         self.plugin_scripts: List[str] = []
+        self.subscribed_generators = [KneeboardGenerator, BriefingGenerator]
```
If possible, don't add any members to this type. This ends up serialized as part of the `Game` so this affects save game compat. It looks like these could just as easily be passed when `notife_subscribed_generators` is called.

(the fact that this gets serialized is probably a bug in its own right, fwiw, but it looks like a fair amount to untangle)

walterroach replied:
Thoughts on extending the pydcs Mission class and wrapping the save method to notify these mission info classes?

Both the kneeboards and the briefing were being generated twice previously, and it looks like the second generation was to fix a bug where the first was occurring.  

Any time we save a mission file, we want to create the briefing and the kneeboards, and we know the required data is guaranteed to be in it's final state.

walterroach replied:
removed the properties that were added.  notify_subscribed_generators (renamed notify_info_generators in 18a93ef) was also added to the Operator class, I imagine the new method won't cause the save compat problems?

DanAlbert replied:
> Thoughts on extending the pydcs Mission class and wrapping the save method to notify these mission info classes?

Sounds good. Fine to do as a follow up if you prefer.

DanAlbert commented on game/operation/operation.py:
```diff
@@ -201,14 +221,14 @@ def generate(self):
         # Generate ground object first
 
-        groundobjectgen = GroundObjectsGenerator(
+        self.groundobjectgen = GroundObjectsGenerator(
```
Same goes for all these.

walterroach replied:
Removed in 18a93ef

DanAlbert commented on gen/briefinggen.py:
```diff
@@ -1,21 +1,28 @@
 from .airsupportgen import AwacsInfo, TankerInfo
 from .armor import JtacInfo
-from .conflictgen import Conflict
+# from .conflictgen import Conflict
```
Remove if unused

walterroach replied:
Removed

DanAlbert commented on game/operation/frontlineattack.py:
```diff
@@ -35,6 +35,7 @@ def prepare(self, terrain: Terrain, is_quick: bool):
-        self.briefinggen.description = "Provide CAS for the ground forces attacking enemy lines. Operation will be considered successful if total number of enemy units will be lower than your own by a factor of 1.5 (i.e. with 12 units from both sides, enemy forces need to be reduced to at least 8), meaning that you (and, probably, your wingmans) should concentrate on destroying the enemy units. T ...
+        ## TODO: What are these for?
+        # self.briefinggen.title = "Frontline CAS"
+        # self.briefinggen.description = "Provide CAS for the ground forces attacking enemy lines. Operation will be considered successful if total number of enemy units will be lower than your own by a factor of 1.5 (i.e. with 12 units from both sides, enemy forces need to be reduced to at least 8), meaning that you (and, probably, your wingmans) should concentrate on destroying the enemy units. ...
```
Just remove them IMO. We have the git history, no need to keep around commented out code.

walterroach replied:
removed

DanAlbert commented on gen/briefinggen.py:
```diff
@@ -24,19 +31,144 @@ class CommInfo:
+        elif self.enemy_zero:
+            return ("The enemy forces have been crushed, we will be able to make significant progress"
+                    f" toward {self.enemy_base.name}.")
+        return ''
```
A minor thing, but try to stay consistent with quote use. The code base is mostly double quotes atm (same goes for docstrings). Exception is is you need to do `"foo said \"bar\""`, using `'foo said "bar"'` is fine then.

https://www.python.org/dev/peps/pep-0008/#string-quotes

walterroach replied:
I heavily rely on Black in my day to day to fix all my single quotes, but it makes a big git mess in this project.  I'll get my formatter configured, that's far more likely to stick than the counting on my habits changing 🙃

walterroach replied:
Changed all single quotes I added to double

DanAlbert commented on gen/briefinggen.py:
```diff
@@ -24,19 +31,144 @@ class CommInfo:
     """
 
-    def __init__(self, mission: Mission) -> None:
+    def __init__(self, mission: Mission, game: 'Game') -> None:
```
with `from __future__ import annotations` you don't need the quotes.

walterroach replied:
I get a circular import exception if I import Game type, even with from __future__ import annotations.  There's probably an underlying issue there, but until it's discovered I'm not sure of another way to type hint it.

DanAlbert replied:
Whoops, also need:

```
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from game import Game
```

at the top. That way it only imports `Game` when it needs it for type checking. It's... kind of gross but it's what we have. (I think when it's just in quotes it's documentation for us but doesn't actually help mypy at all, but not sure)

DanAlbert commented on gen/briefinggen.py:
```diff
@@ -24,19 +31,144 @@ class CommInfo:
+        self.advantage: bool = self.player_base.base.total_armor > self.enemy_base.base.total_armor
+        self.stance: CombatStance = self.player_base.stances[self.enemy_base.id]
+
+    @property
```
All of the properties here could be moved to the template.

walterroach replied:
They can be- but you end up with a long chain of object.object.object.object.object.  I broke them out to help readability, can access the attributes directly if preferred.

DanAlbert commented on gen/briefinggen.py:
```diff
@@ -79,20 +211,13 @@ def add_tanker(self, tanker: TankerInfo) -> None:
-        self.description = ""
-        self.dynamic_runways: List[RunwayData] = []
+        Arguments:
+            frontline {FrontLineInfo} -- [description]
```
I don't think you finished writing this :) Either finish it or remove the `Args` section (we haven't been very diligent at writing them, so either way works for me)

Type info is usually omitted in the argument list since the function declaration already has that covered.

walterroach replied:
fixed

DanAlbert commented on gen/briefinggen.py:
```diff
@@ -102,150 +227,52 @@ def add_dynamic_runway(self, runway: RunwayData) -> None:
+        super().__init__(mission, game)
+        # self.conflict = conflict
+        self.title = ""
+        self.description = ""
```
Aren't these the things you commented out?

DanAlbert commented on gen/kneeboard.py:
```diff
@@ -310,8 +311,8 @@ def format_frequency(self, frequency: RadioFrequency) -> str:
 
-    def __init__(self, mission: Mission) -> None:
-        super().__init__(mission)
+    def __init__(self, mission: Mission, game: 'Game') -> None:
```
Same as before

DanAlbert commented on qt_ui/windows/basemenu/ground_forces/QGroundForcesStrategySelector.py:
```diff
@@ -14,8 +14,8 @@ def __init__(self, cp:ControlPoint, enemy_cp:ControlPoint):
         for i, stance in enumerate(CombatStance):
-            self.addItem(stance.name, userData=stance.value)
-            if self.cp.stances[enemy_cp.id] == stance.value:
+            self.addItem(stance.name, userData=CombatStance(i))
```
Isn't this just `stance`?

walterroach replied:
Yep 😄

DanAlbert reviewed (changes requested):
Mostly good, a few notes though.

I think the only big change desired here to to move _more_ into the template, but this is clearly a step in the right direction so fine by me if you want to do that as a follow up.

walterroach commented:
> I think the only big change desired here to to move _more_ into the template, but this is clearly a step in the right direction so fine by me if you want to do that as a follow up.

I didn't want to spend a lot of time in the template without knowing if we would keep jinja.  The way it is now would be easy to generate a big string in pure Python, and should still be easier to maintain than the existing mega-method.

walterroach committed:
Dan'ts notes

commit 35ab9103cebf823bab85fbb1c9ff4bc2b9c5701a
Author: walterroach <37820425+walterroach@users.noreply.github.com>
Date:   Fri Oct 30 22:25:42 2020 -0500

    more cleanup

commit d3d008bc6b29f328ad48063bd4b0520e1b819d68
Author: walterroach <37820425+walterroach@users.noreply.github.com>
Date:   Fri Oct 30 22:17:59 2020 -0500

    More briefinggen cleanup

commit b2033f091d7191aecefb86fecb5cb060c074706e
Author: walterroach <37820425+walterroach@users.noreply.github.com>
Date:   Fri Oct 30 22:08:48 2020 -0500

    briefinggen cleanup

commit 72ea0196c22d9493df078765800f3fafb9c054dc
Author: walterroach <37820425+walterroach@users.noreply.github.com>
Date:   Fri Oct 30 21:57:52 2020 -0500

    Add notifier method back to Operation

commit efd39a9e03d02b9d581637d0d1c289af68e749c3
Author: walterroach <37820425+walterroach@users.noreply.github.com>
Date:   Fri Oct 30 21:34:37 2020 -0500

    Revert "Move kneeboard and briefing gen to unified"
    Removes properties added to Operation
    This reverts commit 941f2af77074d79b992ae859a0108b67aac5460c.

walterroach committed:
remove useless props

walterroach committed:
added notify call back

DanAlbert commented on game/operation/operation.py:
```diff
@@ -90,8 +90,7 @@ def is_player_attack(self) -> bool:
         self.conflict = conflict
-        self.briefinggen = BriefingGenerator(self.current_mission,
-                                             self.conflict, self.game)
+        # self.briefinggen = BriefingGenerator(self.current_mission, self.game)  Is it safe to remove this, or does it also break save compat?
```
Breaking compat is fine right now. We just want to make sure there's as little as possible left in this class by the time we ship.

In other words, go ahead and remove it.

DanAlbert commented on gen/aircraft.py:
```diff
@@ -272,6 +273,10 @@ def client_units(self) -> List[FlyingUnit]:
+    
+    @property
+    def departure_delay_delta(self) -> timedelta:
+        return timedelta(seconds=self.departure_delay)
```
Good follow up cleanup would be to change the type of the property. I just replaced a whole bunch of ints with timedeltas, but seems I missed this one.

DanAlbert reviewed (approved):
Just a few nits that can be cleaned up after.
